File: GenerateArrows.kt
import java.io.File
import kotlin.random.Random

data class Coordinate(val x: Int, val y: Int)

enum class Direction(val dx: Int, val dy: Int) {
    UP(0, -1),
    DOWN(0, 1),
    LEFT(-1, 0),
    RIGHT(1, 0);

    val opposite: Direction
        get() = when (this) {
            UP -> DOWN
            DOWN -> UP
            LEFT -> RIGHT
            RIGHT -> LEFT
        }
}

data class Arrow(
    val id: Int,
    val segments: List<Coordinate>,
    val direction: Direction,
    val color: Int
)

fun generate(width: Int, height: Int, density: Float, random: Random = Random): List<Arrow> {
    val grid = Array(height) { IntArray(width) { 0 } }
    val arrows = mutableListOf<Arrow>()
    var nextId = 1
    
    val targetCells = (width * height * density).toInt()
    var filledCells = 0
    val colors = listOf(1, 2, 3, 4, 5, 6, 7)

    var attempts = 0
    val maxAttempts = targetCells * 20

    while (filledCells < targetCells && attempts < maxAttempts) {
        attempts++
        
        val edge = random.nextInt(4)
        val spawnDir = when (edge) {
            0 -> Direction.UP    // Spawns at bottom edge, looking UP, pushing backwards means moving DOWN
            1 -> Direction.DOWN  // Spawns at top edge, looking DOWN, pushing backwards means moving UP
            2 -> Direction.LEFT  // Spawns at right edge, looking LEFT, pushing backwards means moving RIGHT
            else -> Direction.RIGHT // Spawns at left edge, looking RIGHT, pushing backwards means moving LEFT
        }

        val headX: Int
        val headY: Int
        when (spawnDir) {
            Direction.UP -> {
                headX = random.nextInt(width)
                headY = height - 1
            }
            Direction.DOWN -> {
                headX = random.nextInt(width)
                headY = 0
            }
            Direction.LEFT -> {
                headX = width - 1
                headY = random.nextInt(height)
            }
            Direction.RIGHT -> {
                headX = 0
                headY = random.nextInt(height)
            }
        }

        if (grid[headY][headX] != 0) continue

        val length = random.nextInt(1, 6)
        
        val segments = mutableListOf<Coordinate>()
        var currentX = headX
        var currentY = headY
        var currentDir = spawnDir.opposite
        
        segments.add(Coordinate(currentX, currentY))

        // We push the head backwards. If the head moves, it traces the body.
        // Actually, the simplest way is to trace the head backward for 'length' steps.
        // Wait, if an arrow has length N, when we push it backward, it traces a path.
        // Let's just generate its final resting shape on the grid.
        
        for (i in 1 until length) {
            if (random.nextFloat() < 0.3f) {
                val turnOptions = listOf(Direction.UP, Direction.DOWN, Direction.LEFT, Direction.RIGHT)
                    .filter { it != currentDir && it != currentDir.opposite }
                currentDir = turnOptions.random(random)
            }

            val nextX = currentX + currentDir.dx
            val nextY = currentY + currentDir.dy

            if (nextX !in 0 until width || nextY !in 0 until height) break
            if (grid[nextY][nextX] != 0) break
            if (segments.contains(Coordinate(nextX, nextY))) break

            currentX = nextX
            currentY = nextY
            segments.add(Coordinate(currentX, currentY))
        }

        val arrowId = nextId++
        for (seg in segments) {
            grid[seg.y][seg.x] = arrowId
            filledCells++
        }
        
        arrows.add(Arrow(id = arrowId, segments = segments, direction = spawnDir, color = colors.random(random)))
    }

    return arrows
}

fun main() {
    val random = Random(42)
    val sb = StringBuilder()
    sb.appendLine("package com.funkyotc.puzzleverse.arrowescape.data")
    sb.appendLine()
    sb.appendLine("import com.funkyotc.puzzleverse.arrowescape.model.Arrow")
    sb.appendLine("import com.funkyotc.puzzleverse.arrowescape.model.Coordinate")
    sb.appendLine("import com.funkyotc.puzzleverse.arrowescape.model.Direction")
    sb.appendLine()
    sb.appendLine("object ArrowEscapePregenerated {")
    
    val sizes = listOf(10, 20, 30)
    val difficulties = listOf("Easy", "Medium", "Hard")
    
    sb.appendLine("    val PUZZLES_BY_DIFFICULTY = mapOf(")
    
    for (i in sizes.indices) {
        val size = sizes[i]
        val diff = difficulties[i]
        
        sb.appendLine("        \"$diff\" to listOf(")
        for (j in 1..5) {
            val arrows = generate(size, size, 0.4f, random)
            sb.appendLine("            listOf(")
            for (arrow in arrows) {
                val segStr = arrow.segments.joinToString(", ") { "Coordinate(${it.x}, ${it.y})" }
                sb.appendLine("                Arrow(${arrow.id}, listOf($segStr), Direction.${arrow.direction}, ${arrow.color}),")
            }
            sb.appendLine("            ),")
        }
        sb.appendLine("        ),")
    }
    
    sb.appendLine("    )")
    sb.appendLine("}")
    
    File("app/src/main/java/com/funkyotc/puzzleverse/arrowescape/data/ArrowEscapePregenerated.kt").apply {
        parentFile.mkdirs()
        writeText(sb.toString())
    }
    println("Generated puzzles.")
}

File: generate_arrows.py
import random
import os

class Direction:
    UP = (0, -1)
    DOWN = (0, 1)
    LEFT = (-1, 0)
    RIGHT = (1, 0)

def opposite(d):
    if d == Direction.UP: return Direction.DOWN
    if d == Direction.DOWN: return Direction.UP
    if d == Direction.LEFT: return Direction.RIGHT
    if d == Direction.RIGHT: return Direction.LEFT

def to_string(d):
    if d == Direction.UP: return "UP"
    if d == Direction.DOWN: return "DOWN"
    if d == Direction.LEFT: return "LEFT"
    if d == Direction.RIGHT: return "RIGHT"

def generate(width, height, density, seed):
    rng = random.Random(seed)
    grid = [[0 for _ in range(width)] for _ in range(height)]
    arrows = []
    next_id = 1
    
    target_cells = int(width * height * density)
    filled_cells = 0
    colors = [1, 2, 3, 4, 5, 6, 7]

    attempts = 0
    max_attempts = target_cells * 20

    while filled_cells < target_cells and attempts < max_attempts:
        attempts += 1
        
        edge = rng.randint(0, 3)
        if edge == 0:
            spawn_dir = Direction.UP
            head_x = rng.randint(0, width - 1)
            head_y = height - 1
        elif edge == 1:
            spawn_dir = Direction.DOWN
            head_x = rng.randint(0, width - 1)
            head_y = 0
        elif edge == 2:
            spawn_dir = Direction.LEFT
            head_x = width - 1
            head_y = rng.randint(0, height - 1)
        else:
            spawn_dir = Direction.RIGHT
            head_x = 0
            head_y = rng.randint(0, height - 1)

        if grid[head_y][head_x] != 0:
            continue

        length = rng.randint(1, 5)
        segments = []
        current_x = head_x
        current_y = head_y
        current_dir = opposite(spawn_dir)
        
        segments.append((current_x, current_y))

        for i in range(1, length):
            if rng.random() < 0.3:
                turn_options = [d for d in [Direction.UP, Direction.DOWN, Direction.LEFT, Direction.RIGHT] if d != current_dir and d != opposite(current_dir)]
                current_dir = rng.choice(turn_options)

            next_x = current_x + current_dir[0]
            next_y = current_y + current_dir[1]

            if next_x < 0 or next_x >= width or next_y < 0 or next_y >= height:
                break
            if grid[next_y][next_x] != 0:
                break
            if (next_x, next_y) in segments:
                break

            current_x = next_x
            current_y = next_y
            segments.append((current_x, current_y))

        arrow_id = next_id
        next_id += 1
        for x, y in segments:
            grid[y][x] = arrow_id
            filled_cells += 1
        
        arrows.append({
            "id": arrow_id,
            "segments": segments,
            "direction": spawn_dir,
            "color": rng.choice(colors)
        })

    return arrows

def main():
    sb = []
    sb.append("package com.funkyotc.puzzleverse.arrowescape.data\n")
    sb.append("import com.funkyotc.puzzleverse.arrowescape.model.Arrow\n")
    sb.append("import com.funkyotc.puzzleverse.arrowescape.model.Coordinate\n")
    sb.append("import com.funkyotc.puzzleverse.arrowescape.model.Direction\n\n")
    sb.append("object ArrowEscapePregenerated {\n")
    
    sizes = [10, 20, 30]
    difficulties = ["Easy", "Medium", "Hard"]
    
    sb.append("    val PUZZLES_BY_DIFFICULTY = mapOf(\n")
    
    for i in range(len(sizes)):
        size = sizes[i]
        diff = difficulties[i]
        
        sb.append(f'        "{diff}" to listOf(\n')
        for j in range(5):
            arrows = generate(size, size, 0.4, 42 + i*10 + j)
            sb.append("            listOf(\n")
            for arrow in arrows:
                seg_str = ", ".join([f"Coordinate({x}, {y})" for x, y in arrow["segments"]])
                dir_str = to_string(arrow["direction"])
                sb.append(f'                Arrow({arrow["id"]}, listOf({seg_str}), Direction.{dir_str}, {arrow["color"]}),\n')
            sb.append("            ),\n")
        sb.append("        ),\n")
    
    sb.append("    )\n")
    sb.append("}\n")
    
    os.makedirs("app/src/main/java/com/funkyotc/puzzleverse/arrowescape/data", exist_ok=True)
    with open("app/src/main/java/com/funkyotc/puzzleverse/arrowescape/data/ArrowEscapePregenerated.kt", "w") as f:
        f.write("".join(sb))
        
    print("Generated puzzles.")

if __name__ == "__main__":
    main()
